### PR TITLE
fix(types): allow no return in asyncData

### PR DESCRIPTION
**Pull request by rchl**
asyncData can be used to, for example decide if the page should
redirect or other things than returning a data object. Having no `void`
return type forces me to have a `return` statement at the end of function
body which looks weird and is unnecessary.

### CODE REVIEW

**Comment by kevinmarrec** [comment]
@rchl AFAIK redirects should be done through `validate`.
And about the type, I'm not sure why I put `undefined` (probably cause it can be).

`asyncData` should aways return something, even if `undefined`, I think.

/cc @Atinux

**Comment by rchl** [comment]
(I think we had that conversation before somewhere but I don't remember previous arguments so I'll try again.)

The thing is, from the code perspective it doesn't make sense to allow one and not the other IMO as one is just implicit return of `undefined` and the other is explicit. So if we allow `undefined` to be returned, we should also allow `void` IMO. See this example:
```js
asyncData(context) {
  if (context.query.foo) {
    return;  // EXPLICIT return without value
  }

  if (context.query.bar) {
    return undefined;  // EXPLICIT return with value
  }

  // IMPLICIT return.
}
```
So all those three in fact return `undefined` but from TS perspective, only `return undefined` is allowed. It's IMO not that common in JS world to exlicitly return `undefined` and it looks weird. Generally people just `return;` which also looks cleaner. And it also looks weird if at the end of the asyncData function I have to put `return undefined` when I could just avoid that line altogether for the same result.

Weird:

```js
asyncData(context) {
  // do something...
  return undefined;
}
```

Nicer:

```js
asyncData(context) {
  // do something...
}
```

**Comment by kevinmarrec** [comment]
@rchl You're right, I'm just wondering if it shouldn't just be `object` or `void`, and never `undefined`

**Comment by rchl** [comment]
Void would imo be better and enough. Since you can still afford making breaking changes as package is at 0.x.x version, I'd vote for that.

**Comment by kevinmarrec** [comment]
@rchl Right, please update your PR then by just replacing `undefined` by `void` :)

**Comment by rchl** [comment]
Actually I think it should also have Promise<void> and Promise<object> return types...

**Comment by kevinmarrec** [comment]
That's right, but after all, `Promise` is an object ahah 😁

Still, sounds good to me

File: packages/types/app/vue.d.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { Context, Middleware, Transition, NuxtApp } from './index'
 
 declare module 'vue/types/options' {
   interface ComponentOptions<V extends Vue> {
-    asyncData?(ctx: Context): object | undefined
+    asyncData?(ctx: Context): Promise<object> | Promise<void> | object | void
```
Comment (kevinmarrec) [inline packages/types/app/vue.d.ts]:
```suggestion
    asyncData?(ctx: Context): Promise<object | void> | object | void
```
should also work

**Pull request merged**